pkg/storage/etcd3/store.go

```diff
@@ -0,0 +1,310 @@
+	for _, member := range resp.Members {
+		mlist = append(mlist, member.ClientURLs...)
+	}
+	return mlist
+}
+
+func (s *store) Codec() runtime.Codec {
+	return s.codec
+}
+
+func (s *store) Versioner() storage.Versioner {
+	return s.versioner
+}
+
+func (s *store) Set(ctx context.Context, key string, obj, out runtime.Object, ttl uint64) error {
+	rev, err := extractAndResetRevision(s.versioner, obj)
+	if err != nil {
+		return err
+	}
+	data, err := runtime.Encode(s.codec, obj)
+	if err != nil {
+		return err
+	}
+	key = keyWithPrefix(s.pathPrefix, key)
+
+	// Note that if rev = 0, it means the key doesn't exist. In such case, it's equivalent to create.
+	txnResp, err := s.client.KV.Txn(ctx).If(
+		clientv3.Compare(clientv3.ModifiedRevision(key), "=", rev),
+	).Then(
+		clientv3.OpPut(key, string(data)),
+	).Commit()
+	if err != nil {
+		return err
+	}
+	if !txnResp.Succeeded {
+		return storage.NewResourceVersionConflictsError(key, rev)
+	}
+
+	if out != nil {
+		putResp := txnResp.Responses[0].GetResponsePut()
+		return decode(s.codec, s.versioner, data, out, putResp.Header.Revision)
+	}
+
+	return nil
+}
+
+func (s *store) Get(ctx context.Context, key string, out runtime.Object, ignoreNotFound bool) error {
+	key = keyWithPrefix(s.pathPrefix, key)
+	getResp, err := s.client.KV.Get(ctx, key)
+	if err != nil {
+		return err
+	}
+
+	if len(getResp.Kvs) == 0 {
+		if ignoreNotFound {
+			return runtime.SetZeroValue(out)
+		}
+		return storage.NewKeyNotFoundError(key, 0)
+	}
+
+	kv := getResp.Kvs[0]
+	return decode(s.codec, s.versioner, kv.Value, out, kv.ModRevision)
+}
+
+func (s *store) Create(ctx context.Context, key string, obj, out runtime.Object, ttl uint64) error {
+	if version, err := s.versioner.ObjectResourceVersion(obj); err == nil && version != 0 {
+		return errors.New("resourceVersion should not be set on objects to be created")
+	}
+	data, err := runtime.Encode(s.codec, obj)
+	if err != nil {
+		return err
+	}
+	key = keyWithPrefix(s.pathPrefix, key)
+
+	txnResp, err := s.client.KV.Txn(ctx).If(
+		notFound(key),
+	).Then(
+		clientv3.OpPut(key, string(data)),
+	).Commit()
+	if err != nil {
+		return err
+	}
+	if !txnResp.Succeeded {
+		return storage.NewKeyExistsError(key, 0)
+	}
+
+	if out != nil {
+		putResp := txnResp.Responses[0].GetResponsePut()
+		return decode(s.codec, s.versioner, data, out, putResp.Header.Revision)
+	}
+	return nil
+}
+
+func (s *store) Delete(ctx context.Context, key string, out runtime.Object) error {
+	if _, err := conversion.EnforcePtr(out); err != nil {
+		panic("unable to convert output object to pointer")
+	}
+
+	key = keyWithPrefix(s.pathPrefix, key)
+	// We need to do get and delete in single transaction in order to
+	// know the value and revision before deleting it.
+	txnResp, err := s.client.KV.Txn(ctx).If().Then(
+		clientv3.OpGet(key),
+		clientv3.OpDelete(key),
+	).Commit()
+	if err != nil {
+		return err
+	}
+	getResp := txnResp.Responses[0].GetResponseRange()
+	if len(getResp.Kvs) == 0 {
```

Comment:
in some case we may want to ignore the not found error if we are deleting not exist key.